Update reference for target_feature_11.

veluca93 opened the pull request:
Tracking issue: https://github.com/rust-lang/rust/issues/69098
Stabilization proposal: https://github.com/rust-lang/rust/pull/134090

veluca93 committed:
Update reference for target_feature_11.

ehuss commented:
Note there was a previous PR at https://github.com/rust-lang/reference/pull/1181 that includes some comments of things that were missing or confusing. For example https://github.com/rust-lang/reference/pull/1181#issuecomment-1468707948, contained some changes that happened over time that hadn't yet been captured.

veluca93 commented:
> Note there was a previous PR at #1181 that includes some comments of things that were missing or confusing. For example [#1181 (comment)](https://github.com/rust-lang/reference/pull/1181#issuecomment-1468707948), contained some changes that happened over time that hadn't yet been captured.

I was not aware of the existing PR!

I believe I have integrated the main comments from that PR, PTAL :-)

chorman0773 reviewed (approved):
Content looks to be correct (verified by short code tests) and I don't have any editorial concerns.

ehuss commented:
Am I correct in understanding that this is a new "unsafe superpower"? That is, do we need to add a new item to `unsafety.md` that defines something along the lines of calling a safe function with an unmatched `target_feature`?

veluca93 commented on src/attributes/codegen.md:
```diff
@@ -69,19 +69,45 @@ features. It uses the [_MetaListNameValueStr_] syntax with a single key of
+r[attributes.codegen.target_feature.safety-restrictions]
+Because of this, on many platforms the following restrictions apply:
+
+- `#[target_feature]` functions (and closures that inherit the attribute)
+  can only be safely called within caller that enable all the `target_feature`s
+  that the callee enables.
```
This does include implicitly enabled features on either side.
Although of course implicitly enabled features on the callee are enabled on the caller too if all the explicitly enabled features in the callee are enabled...

Do you think that is something that should be clarified?

ehuss replied:
Ah, ok. It looks like this changed in https://github.com/rust-lang/rust/pull/128221 since the last time I looked at this.

I'd like to do a quick editorial pass, and I'll try to clarify this point. I'll try to get to that today.

veluca93 commented:
> Am I correct in understanding that this is a new "unsafe superpower"? That is, do we need to add a new item to `unsafety.md` that defines something along the lines of calling a safe function with an unmatched `target_feature`?

I think this is technically true, yes (in some sense it is not new, because tf functions were unsafe before, but it's probably worth mentioning explicitly)

ehuss committed:
Add target_feature to unsafety list

ehuss committed:
Reward the `unsafe` block to be more general

The `unsafe` block enables several abilities beyond just calling unsafe
functions and dereferencing pointers. Link to the actual list of things
you can do.

ehuss committed:
Clean up the markdown

ehuss committed:
Rework the target_fature safety restrictions

This does some rewording to try to make things a little more explicit
and clearer:

- Moves the `Fn` traits to a separate rule, it wasn't directly related
  to what is safe and not safe.
- Moves the allowed positions into a separate rule, and adds some links
  to those. I dropped the "other special functions" because it is not
  defined anywhere.
- Add an example.
- Add a remark about implicit features.
- Don't say "on many platforms", and be more explicit about how this
  is overridden.
- Various rewording for clarity.
- Remove word wrapping.

ehuss reviewed (approved):
Looks like https://github.com/rust-lang/rust/pull/134090 has merged, thanks for working on this!
